src/gui/MainWindow.cpp

```diff
@@ -1281,40 +1279,46 @@ void MainWindow::updateTrayIcon()
             menu->addAction(m_ui->actionLockDatabases);
 
 #ifdef Q_OS_MACOS
-            QAction* actionQuit = new QAction(tr("Quit KeePassXC"), menu);
-            menu->addAction(actionQuit);
-
+            auto actionQuit = new QAction(tr("Quit KeePassXC"), menu);
             connect(actionQuit, SIGNAL(triggered()), SLOT(appExit()));
+            menu->addAction(actionQuit);
 #else
             menu->addAction(m_ui->actionQuit);
-
 #endif
+            m_trayIcon->setContextMenu(menu);
+
             connect(m_trayIcon,
                     SIGNAL(activated(QSystemTrayIcon::ActivationReason)),
                     SLOT(trayIconTriggered(QSystemTrayIcon::ActivationReason)));
             connect(actionToggle, SIGNAL(triggered()), SLOT(toggleWindow()));
-
-            m_trayIcon->setContextMenu(menu);
-
-            m_trayIcon->setIcon(resources()->trayIcon());
-            m_trayIcon->show();
         }
 
-        if (m_ui->tabWidget->count() == 0) {
-            m_trayIcon->setIcon(resources()->trayIcon());
-        } else if (m_ui->tabWidget->hasLockableDatabases()) {
+        if (m_ui->tabWidget->hasLockableDatabases()) {
             m_trayIcon->setIcon(resources()->trayIconUnlocked());
         } else {
             m_trayIcon->setIcon(resources()->trayIconLocked());
         }
-    } else {
-        QApplication::setQuitOnLastWindowClosed(true);
 
+        m_trayIcon->setToolTip(windowTitle().replace("[*]", isWindowModified() ? "*" : ""));
```

Comment:
Can we somehow construct the tooltip from the original raw title + an asterisk? Updating the existing title will replace any asterisks appearing in the database name. We should never read anything back from the UI.

Reply:
That is the explicit placeholder for the modified indicator. It's a stupid qt convention and they keep the placeholder intact for some reason.

Reply:
Hmm. I really don't like it, but so be it.